src/runtime/stream.rs

```diff
@@ -0,0 +1,293 @@
+    #[cfg(feature = "tokio-runtime")]
+    Tokio(tokio::net::TcpStream),
+
+    /// Wrapper around `async_std::net::TcpStream`.
+    #[cfg(feature = "async-std-runtime")]
+    AsyncStd(async_std::net::TcpStream),
+}
+
+#[async_trait::async_trait]
+trait AsyncConnect
+where
+    Self: Sized,
+{
+    async fn connect(address: &SocketAddr) -> Result<Self>;
+
+    async fn connect_timeout(address: &SocketAddr, timeout: Duration) -> Result<Self>;
+}
+
+#[cfg(feature = "tokio-runtime")]
+#[async_trait::async_trait]
+impl AsyncConnect for tokio::net::TcpStream {
+    async fn connect(address: &SocketAddr) -> Result<Self> {
+        let stream = Self::connect(address).await?;
+        Ok(stream)
+    }
+
+    async fn connect_timeout(address: &SocketAddr, timeout: Duration) -> Result<Self> {
+        let stream = Self::connect_timeout(address, timeout).await?;
+        Ok(stream)
+    }
+}
+
+#[cfg(feature = "async-std-runtime")]
+#[async_trait::async_trait]
+impl AsyncConnect for async_std::net::TcpStream {
+    async fn connect(address: &SocketAddr) -> Result<Self> {
+        let stream = Self::connect(address).await?;
+        Ok(stream)
+    }
+
+    async fn connect_timeout(address: &SocketAddr, timeout: Duration) -> Result<Self> {
+        let stream = Self::connect_timeout(address, timeout).await?;
+        Ok(stream)
+    }
+}
+
+async fn try_connect<C: AsyncConnect>(address: &SocketAddr, timeout: Duration) -> Result<C> {
+    // The URI options spec requires that the default connect timeout is 10 seconds, but that 0
+    // should indicate no timeout.
+    let stream = if timeout == Duration::from_secs(0) {
+        C::connect(address).await?
+    } else {
+        C::connect_timeout(address, timeout).await?
+    };
+
+    Ok(stream)
+}
+
+async fn connect_stream<C: AsyncConnect>(
+    address: &StreamAddress,
+    connect_timeout: Option<Duration>,
+) -> Result<C> {
+    let timeout = connect_timeout.unwrap_or(DEFAULT_CONNECT_TIMEOUT);
+
+    let mut socket_addrs: Vec<_> = address.to_socket_addrs()?.collect();
+
+    if socket_addrs.is_empty() {
+        return Err(ErrorKind::NoDnsResults(address.clone()).into());
+    }
+
+    // After considering various approaches, we decided to do what other drivers do, namely try each
+    // of the addresses in sequence with a preference for IPv4.
+    socket_addrs.sort_by_key(|addr| if addr.is_ipv4() { 0 } else { 1 });
+
+    let mut connect_error = None;
+
+    for address in &socket_addrs {
+        connect_error = match try_connect(address, timeout).await {
+            Ok(stream) => return Ok(stream),
+            Err(err) => Some(err),
+        };
+    }
+
+    Err(connect_error.unwrap_or_else(|| ErrorKind::NoDnsResults(address.clone()).into()))
+}
+
+impl AsyncStream {
+    /// Creates a new Tokio TCP stream connected to the server as specified by `options`.
+    #[cfg(feature = "tokio-runtime")]
+    pub(super) async fn connect_tokio(options: StreamOptions) -> Result<Self> {
```

Comment:
This is a copy of the sync `Stream::connect` with the only changes being using tokio's `TcpStream` and tokio_rustls's `TlsStream` instead of those from `std` and `rustls` respectively and making the function async and awaiting where necessary.